### PR TITLE
chore: Improve ability to override Checkbox and Radio styles

### PR DESCRIPTION
**Pull request by behowell**
## Previous Behavior

The styles for Checkbox and Radio use complex state and sibling selectors, which can be tricky to override correctly. See #25483 for more details.

## New Behavior

For any color set via a selector in Checkbox and Radio, set the value of a CSS variable rather than the CSS property directly. Then in that slot's base styles (no selectors), apply the current value of the CSS variable.

This eliminates any specificity issues when overriding the CSS properties for any given slot. Users can now override the values for any slot using any selector they would like. For example:

```jsx
const useStyles = makeStyles({
  labelUnchecked: {
    color: tokens.colorPaletteRedForeground1,
  },
  labelChecked: {
    color: tokens.colorPaletteGreenForeground1,
  },
  indicatorUnchecked: {
    backgroundColor: tokens.colorPaletteRedBackground1,
    ...shorthands.borderColor(tokens.colorPaletteRedBorder1),
  },
  indicatorChecked: {
    color: tokens.colorPaletteGreenForeground1,
    backgroundColor: tokens.colorPaletteGreenBackground1,
    ...shorthands.borderColor(tokens.colorPaletteGreenBorder1),
  },
});

export const ColorfulCheckbox = (props: CheckboxProps) => {
  const [checked, setChecked] = useControllableState({
    defaultState: props.defaultChecked,
    state: props.checked,
    initialState: false,
  });

  const styles = useStyles();

  return (
    <Checkbox
      checked={checked}
      onChange={(_ev, data) => setChecked(data.checked)}
      label={resolveShorthand(props.label, {
        required: true,
        defaultProps: { className: mergeClasses(checked ? styles.labelChecked : styles.labelUnchecked) },
      })}
      indicator={resolveShorthand(props.indicator, {
        required: true,
        defaultProps: { className: mergeClasses(checked ? styles.indicatorChecked : styles.indicatorUnchecked) },
      })}
    />
  );
};
```

image

## Related Issue(s)

* #25483
* This PR is an alternative solution to #26247. Only one of these two PRs will be merged.

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/Avatar-mount_base.html">1513</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/Avatar-mount.html">1537</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/Button-mount_base.html">1121</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/Button-mount.html">1111</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/FluentProvider-mount_base.html">1908</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/FluentProvider-mount.html">1850</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/FluentProviderWithTheme-mount_base.html">746</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/FluentProviderWithTheme-mount.html">755</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">708</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">728</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">749</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">741</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/MakeStyles-mount_base.html">2278</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/MakeStyles-mount.html">2263</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/Persona-mount_base.html">3582</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/Persona-mount.html">3550</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/SpinButton-mount_base.html">3008</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26004/perf-test-react-components/SpinButton-mount.html">2931</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                |     Baseline (minified/GZIP) |                           PR |                                                                                                                                                                                                      Change |
| :----------------------------------------------------------------------------------------------- | ---------------------------: | ---------------------------: | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-checkbox</samp> <br /> <abbr title='bundle-size/Checkbox.fixture.js'>Checkbox</abbr> |    `28.65 kB`<br />`9.09 kB` |  `29.944 kB`<br />`9.153 kB` |  `1.294 kB` <br />`63 B`  |
| <samp>react-radio</samp> <br /> <abbr title='bundle-size/Radio.fixture.js'>Radio</abbr>          | `31.827 kB`<br />`10.316 kB` | `33.699 kB`<br />`10.495 kB` | `1.872 kB` <br />`179 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr>              |  `59.288 kB`<br />`16.435 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `186.264 kB`<br />`52.303 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |   `33.75 kB`<br />`11.101 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                                         |    `5.857 kB`<br />`1.978 kB` |
| <samp>react-radio</samp> <br /> <abbr title='bundle-size/RadioGroup.fixture.js'>RadioGroup</abbr>                                                                             |    `14.304 kB`<br />`5.72 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/983c8a8477ffb675d5c756b5407d926bb90a5ad7'>983c8a8477ffb675d5c756b5407d926bb90a5ad7</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluentuiv9"/>

File: packages/react-components/react-checkbox/src/components/Checkbox/useCheckboxStyles.ts [hunk 1]
```diff
@@ -165,24 +177,24 @@ const useIndicatorBaseClassName = makeResetStyles({
...
+  fontSize: `calc(${values.indicatorSize} - 4px)`,
+  height: values.indicatorSize,
+  width: values.indicatorSize,
 
-const useIndicatorStyles = makeStyles({
-  large: {
-    fontSize: '16px',
-    height: indicatorSizeLarge,
-    width: indicatorSizeLarge,
+  '> *': {
```
Comment (layershifter) [inline packages/react-components/react-checkbox/src/components/Checkbox/useCheckboxStyles.ts:192]: This is an super expensive selector as it matches all elements on a page first.

Comment (behowell) [reply packages/react-components/react-checkbox/src/components/Checkbox/useCheckboxStyles.ts:192]: This `> *` selector already existed in the Checkbox styles; it's just being refactored in this PR:
```
[`:not(:checked):not(:indeterminate) ~ .${checkboxClassNames.indicator} > *`]
```

 I made a separate PR (unrelated to this one) to remove it, and could update this PR once it's in:
* #26248

Comment (behowell) [reply packages/react-components/react-checkbox/src/components/Checkbox/useCheckboxStyles.ts]: Removed

**Comment by behowell** [comment]
After discussing with the team, we decided to go with #26247 instead.

**Pull request closed without merging**